runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java

```diff
@@ -367,7 +406,22 @@ public void visitPrimitiveTransform(TransformHierarchy.Node node) {
       LOG.info("Evaluating {}", transform);
       AppliedPTransform<?, ?, ?> appliedTransform = node.toAppliedPTransform();
       ctxt.setCurrentTransform(appliedTransform);
-      evaluator.evaluate(transform, ctxt);
+      // by default, we don't cache
+      boolean cacheHint = false;
+      // if the output of the node (aka PTransform) is already known in the cache candidates map
+      // (initialized with inputs), and it appears more than one time, then we enable caching
+
+      // considering the node output for caching
+      for (TaggedPValue output : node.getOutputs()) {
+        PValue value = output.getValue();
+        if ((value instanceof PCollection)
+            && cacheCandidates.containsKey(value)
+            && cacheCandidates.get(value) > 1) {
+          cacheHint = true;
+          break;
+        }
+      }
+      evaluator.evaluate(transform, ctxt, cacheHint);
```

Comment:
Can `cacheCandidates` be moved to `EvaluationContext`? This will save on changing `evaluate` signature and you can move the `cacheHint` logic to `EvaluationContext` (add a method like `shouldCache(transform)`)
This way you can also save on changing `EvaluationContext#borrowDataset` signature since evaluation context will be able to ask itself `shouldCache` instead of getting this information via the cacheHint argument.